source/Engine.cpp

```diff
@@ -1205,6 +1205,62 @@ void Engine::BreakTargeting(const Government *gov)
 
 
 
+unsigned Engine::FleetPlacementLimit(const LimitedEvents<Fleet> &fleet, unsigned frames, bool requireGovernment)
+{
+	if(requireGovernment && !fleet.Get()->GetGovernment())
+		// Fleet has no government, but caller required one.
+		return 0;
+	else if(frames && Random::Int(fleet.Period()) >= frames)
+		// It is not yet time to place this fleet.
+		return 0;
+	else if(frames && !fleet.HasLimit())
+		// This is not an initalCount spawn, and the fleet is unlimited.
+		return numeric_limits<int>::max();
+	else if(!frames && fleet.InitialCount() <= 0)
+		// During an initialCount spawn, if the initialCount is 0, there's nothing to spawn.
+		return false;
+
+	int count = CountFleetsWithId(fleet.Id());
+	int maximum = frames ? fleet.Limit() : fleet.InitialCount();
+	return static_cast<unsigned>(max<int>(0, maximum - count));
+}
+
+
+
+unsigned Engine::CountFleetsWithId(const string &id)
+{
+	if(id.empty())
+		return 0;
+	auto range = limitedFleets.equal_range(id);
+	unsigned count = 0;
+//	using iter = std::unordered_multimap<std::string,std::weak_ptr<std::string>>::const_iterator;
+	for(auto it = range.first; it != range.second; )
+//	for(iter it = range.first; it != range.second; )
+		if(it->second.expired())
+			it = limitedFleets.erase(it);
```

Comment:
This function shouldn't count *and* erase. Those two tasks should be separated in different functions.

Reply:
That part of the code is gone now.